File: zig/math/sieve.zig
// Print the primes below 100, sieved over a fixed array of flags. (written against Zig 0.16)
// Build: zig build-exe sieve.zig

const std = @import("std");

const limit = 100;

pub fn main() void {
    var is_prime = [_]bool{true} ** limit;
    is_prime[0] = false;
    is_prime[1] = false;

    var n: usize = 2;
    while (n * n < limit) : (n += 1) {
        if (!is_prime[n]) continue;
        var multiple = n * n;
        while (multiple < limit) : (multiple += n) {
            is_prime[multiple] = false;
        }
    }

    var first = true;
    for (is_prime, 0..) |prime, value| {
        if (!prime) continue;
        if (!first) std.debug.print(" ", .{});
        std.debug.print("{d}", .{value});
        first = false;
    }
    std.debug.print("\n", .{});
}

File: c/math/sieve.c
/* Print the primes below 100, sieved over a fixed array of flags on the stack. */
/* Build: cc -o sieve sieve.c */

#include <stdio.h>

#define LIMIT 100

int main(void)
{
    int is_prime[LIMIT];
    int n;
    int first = 1;

    for (n = 0; n < LIMIT; n++) {
        is_prime[n] = n >= 2;
    }

    for (n = 2; n * n < LIMIT; n++) {
        if (is_prime[n]) {
            int multiple;

            for (multiple = n * n; multiple < LIMIT; multiple += n) {
                is_prime[multiple] = 0;
            }
        }
    }

    for (n = 0; n < LIMIT; n++) {
        if (is_prime[n]) {
            printf("%s%d", first ? "" : " ", n);
            first = 0;
        }
    }
    putchar('\n');
    return 0;
}
